framework/src/main/java/org/checkerframework/framework/type/AbstractViewpointAdapter.java

```diff
@@ -0,0 +1,492 @@
+
+            // Combine component type recursively and sets combined component type
+            AnnotatedTypeMirror compo = aat.getComponentType();
+            // Recursively call itself first on the component type
+            AnnotatedTypeMirror combinedCompoType =
+                    combineAnnotationWithType(receiverAnnotation, compo);
+            aat.setComponentType(combinedCompoType);
+
+            return aat;
+        } else if (declared.getKind() == TypeKind.WILDCARD) {
+            AnnotatedWildcardType awt = (AnnotatedWildcardType) declared.shallowCopy();
+
+            Map<AnnotatedTypeMirror, AnnotatedTypeMirror> mapping = new HashMap<>();
+
+            // There is no main modifier for a wildcard
+
+            // Adapt extend
+            AnnotatedTypeMirror extend = awt.getExtendsBound();
+            if (extend != null) {
+                // Recursively adapt the extends bound of this awt
+                AnnotatedTypeMirror combinedExtend =
+                        combineAnnotationWithType(receiverAnnotation, extend);
+                mapping.put(extend, combinedExtend);
+            }
+
+            // Adapt super
+            AnnotatedTypeMirror zuper = awt.getSuperBound();
+            if (zuper != null) {
+                // Recursively adapt the lower bound of this awt
+                AnnotatedTypeMirror combinedZuper =
+                        combineAnnotationWithType(receiverAnnotation, zuper);
+                mapping.put(zuper, combinedZuper);
+            }
+
+            AnnotatedTypeMirror result = AnnotatedTypeReplacer.replace(awt, mapping);
+
+            return result;
+        } else if (declared.getKind() == TypeKind.NULL) {
+            AnnotatedNullType ant = (AnnotatedNullType) declared.shallowCopy(true);
+            AnnotationMirror declaredAnnotation = extractAnnotationMirror(ant);
+            AnnotationMirror resultAnnotation =
+                    combineAnnotationWithAnnotation(receiverAnnotation, declaredAnnotation);
+            ant.replaceAnnotation(resultAnnotation);
+            return ant;
+        } else {
+            ErrorReporter.errorAbort(
+                    "ViewpointAdaptor::combineAnnotationWithType: Unknown decl: "
+                            + declared
+                            + " of kind: "
+                            + declared.getKind());
+            return null;
+        }
+    }
+
+    /**
+     * Viewpoint adapt declared modifier to receiver modifier.
+     *
+     * @param receiverAnnotation receiver modifier
+     * @param declaredAnnotation declared modifier
+     * @return result modifier after viewpoint adaptation
+     */
+    @SideEffectFree
+    protected abstract AnnotationMirror combineAnnotationWithAnnotation(
+            AnnotationMirror receiverAnnotation, AnnotationMirror declaredAnnotation);
+
+    /**
+     * If rhs is type variable use whose type arguments should be inferred from receiver - lhs, this
```

Comment:
receiver subtract lhs? can you clarify "receiver - lhs"?

Reply:
Done